tests/optimize/test_qnspsa.py

```diff
@@ -0,0 +1,311 @@
+    qnode_finite_diff expression:
+    qnode(params + finite_diff_step * direction) - qnode(params - finite_diff_step * direction)
+    """
+    direction = grad_dirs[0]
+    qnode_finite_diff = (
+        -np.sin(params[0][0]) * np.sin(finite_diff_step * direction[0][0])
+        + np.sin(params[0][1]) * np.sin(finite_diff_step * direction[0][1])
+        + (
+            np.cos(
+                params[0][0]
+                + params[0][1]
+                + finite_diff_step * direction[0][0]
+                + finite_diff_step * direction[0][1]
+            )
+            + np.cos(
+                params[0][0]
+                + finite_diff_step * direction[0][0]
+                - params[0][1]
+                - finite_diff_step * direction[0][1]
+            )
+            - np.cos(
+                params[0][0]
+                + params[0][1]
+                - finite_diff_step * direction[0][0]
+                - finite_diff_step * direction[0][1]
+            )
+            - np.cos(
+                params[0][0]
+                - finite_diff_step * direction[0][0]
+                - params[0][1]
+                + finite_diff_step * direction[0][1]
+            )
+        )
+        / 4
+    )
+    return qnode_finite_diff / (2 * finite_diff_step) * direction
+
+
+def get_metric_from_single_input_qnode(params, finite_diff_step, tensor_dirs):
+    """Compute the expected raw metric tensor from analytical state overlap expression."""
+    dir_vec1 = tensor_dirs[0]
+    dir_vec2 = tensor_dirs[1]
+    dir1 = dir_vec1.reshape(params.shape)
+    dir2 = dir_vec2.reshape(params.shape)
+    perturb1 = dir1 * finite_diff_step
+    perturb2 = dir2 * finite_diff_step
+
+    def get_state_overlap(params1, params2):
+        # analytically computed state overlap between two parameterized ansatzes
+        # with input params1 and params2
+        return (
+            np.cos(params1[0][0] / 2) * np.cos(params2[0][0] / 2)
+            + np.sin(params1[0][0] / 2)
+            * np.sin(params2[0][0] / 2)
+            * (
+                np.sin(params1[0][1] / 2) * np.sin(params2[0][1] / 2)
+                + np.cos(params1[0][1] / 2) * np.cos(params2[0][1] / 2)
+            )
+        ) ** 2
+
+    tensor_finite_diff = (
+        get_state_overlap(params, params + perturb1 + perturb2)
+        - get_state_overlap(params, params + perturb1)
+        - get_state_overlap(params, params - perturb1 + perturb2)
+        + get_state_overlap(params, params - perturb1)
+    )
+    metric_tensor_expected = (
+        -(np.tensordot(dir_vec1, dir_vec2, axes=0) + np.tensordot(dir_vec2, dir_vec1, axes=0))
+        * tensor_finite_diff
+        / (8 * finite_diff_step * finite_diff_step)
+    )
+    return metric_tensor_expected
+
+
+class TestQNSPSAOptimizer:
+    @pytest.mark.parametrize("seed", [1, 151, 1231])
+    @pytest.mark.parametrize("finite_diff_step", [1e-3, 1e-2, 1e-1])
```

Comment:
It seems that all functions in this class have a `seed` and a `finite_diff_step` argument which is parameterized. If these values are the same for each function, then one can actually just parameterize the whole class. Something like: 

```python
@pytest.mark.parameterize("seed", [1, 151, 1231])
class TestQNSPSAOptimizer: 
    ...
```